Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -122,6 +122,14 @@ public final class SwiftLanguageServer: ToolchainLanguageServer {
   var keys: sourcekitd_keys { return sourcekitd.keys }
   var requests: sourcekitd_requests { return sourcekitd.requests }
   var values: sourcekitd_values { return sourcekitd.values }
+
+  var enablePublishDiagnostics: Bool {
+    // Since LSP 3.17.0, diagnostics can be reported through pull-based requests,
+    // in addition to the existing push-based publish notifications.
+    // If the client supports pull diagnostics, we report the capability
+    // and we should disable the publish notifications to avoid double-reporting.
+    return clientCapabilities.textDocument?.diagnostic == nil
```

Comment:
Should this rather be testing `clientCapabilities.textDocument?.diagnostic?.dynamicRegistration != true`? I'm not sure if we would receive the pull requests if dynamic registration is not supported by the client.

Reply:
I think this is the right approach. Correct me if you think I’m wrong but checking for `dynamicRegistration` would only check if the client supports dynamic registration of the capability. If, for some reason, it supported registration of the capability at server initialization but not dynamically after the fact, checking for `dynamicRegistration` wouldn’t catch that.

Reply:
We have the same understanding of `dynamicRegistration`. My read of `SourceKitServer` is that it doesn't report supporting pull diagnostics at initialization. Later, when it is asked about a swift file, it starts `SwiftLanguageServer`, initializes it, sees that it reports supporting pull diagnostics and at that point dynamically registers support for pull diagnostics to the LSP client. So, we depend on `dynamicRegistration` being supported.

Or maybe we should unconditionally report support for pull diagnostics during `SourceKitServer`'s initialization, avoiding dynamic registration? If our underlying language server doesn't support the request, `SourceKitServer` will return no diagnostics so I think there's no harm done.

Reply:
That shouldn't be a problem. `SourceKitServer.initialize()` can unconditionally report support for pull diagnostics, but `SwiftLanguageServer`'s logic is based on the `ClientCapabilities` coming from the LSP Client, so it will only disable publish diagnostics if the LSP client supports pull diagnostics.

Reply:
Ah sorry I'm getting mixed up between the two language servers

Reply:
@ahoppen I'm almost done implementing your suggestion but now that it's taking shape, I'm not sure this is the right approach. It adds some complexity for determining if we should or not publish diagnostics, and it requires dynamic capability registration.

I'm thinking a better approach would be to stop publishing diagnostics in the language-specific server as soon as we receive a first pull diagnostics request, so we have proof that the client is using them. We could publish an empty diagnostics array at that point to clear up any previously published diagnostics, then set a flag to not publish any further diagnostics and only respond to pull requests.

I also think it makes sense to unconditionally reporting pull diagnostics as supported in the SourceKitServer (independent of language-specific servers), rather than using dynamic registration when one language-specific server comes online and requires it. We have a well-defined response if we get a pull diagnostics request and the language-specific server doesn't support it (return an empty diagnostics array), and it won't stop the client from responding to publish diagnostics messages by language-specific servers who do not support pull diagnostics. Thoughts?

Reply:
Honestly, what I really want, would be to see how other language servers handle this. But I haven’t found a single major LSP server that supports both `publishDiagnsotics` and `textDocument/diagnostic`. Also, what surprises me, is that I can’t find a PR or issue on https://github.com/microsoft/language-server-protocol that mentions `textDocument/diagnostics`, which kind of surprises me because I thought there would be some kind of discussion about these features. 

Anyway, given the lack of precedence or clarification, I think we really need to make up our own and out of all the alternatives we have considered, I think not sending `publishDiagnostics` notifications if both the server and the client support “Pull Diagnostics” still seems like the cleanest design to me.

---

> It adds some complexity for determining if we should or not publish diagnostics, and it requires dynamic capability registration.

What’s the added complexity here? I think passing down the `CapabilityRegistry` instead of `ClientCapabilities` as you do it in your PR works rather nicely. I haven’t looked at it in too much detail though.

> I'm thinking a better approach would be to stop publishing diagnostics in the language-specific server as soon as we receive a first pull diagnostics request, so we have proof that the client is using them. 

Hmm, I’m not a huge fan because this means that sending a request silently changes the entire behavior of SourceKit-LSP, which is very confusing IMO. I think whether `publishDiagnostic` notifications should be determined when the client and server agree on capabilities.

> I also think it makes sense to unconditionally reporting pull diagnostics as supported in the SourceKitServer (independent of language-specific servers), rather than using dynamic registration when one language-specific server comes online and requires it. We have a well-defined response if we get a pull diagnostics request and the language-specific server doesn't support it (return an empty diagnostics array), and it won't stop the client from responding to publish diagnostics messages by language-specific servers who do not support pull diagnostics. Thoughts?

But wouldn’t that leave us with the following problem: `clangd` publishes its diagnostics using `publishDiagnostics`. Then the client requests diagnostics using `textDocument/diagnostics` and SourceKit-LSP returns an empty array, clearing any diagnostics displayed on the client.

Reply:
Oh I think that your last point reveals our different assumptions. My assumption is that the namespacing of pull and publish diagnostics is separate and they don't interact because that would be hard to manage, but I'm still looking for some confirmation of this (specs or reference implementations).

Reply:
Oh, yes. If they are indeed in different namespaces, that would certainly change things.

Reply:
Let's see if this gets us more info: https://github.com/microsoft/language-server-protocol/issues/1743

Reply:
In the meanwhile, I finished and pushed your suggested implementation, so we can make some progress. It's ready to review now.